nupic/research/frameworks/continuous_learning/multihead/multihead.py

```diff
@@ -0,0 +1,205 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+Functions for training and testing a network with single/multiple heads in a continuous
+learning framework
+"""
+
+import torch
+import torch.nn.functional as F
+import torch.optim as optim
+from tqdm import tqdm
+
+from nupic.research.frameworks.continuous_learning.multihead.subdataset import (
+    get_datasets,
+)
+
+SEED = 42
+LEARNING_RATE = 0.001
+EPOCHS = 30
+TRAIN_BATCH_SIZE = 64
+TEST_BATCH_SIZE = 1000
+
+
+def train_head(model, loader, optimizer, criterion, device, active_classes=None,
+               post_batch_callback=None):
+    """
+    Train the model on a single task using given dataset loader and set of active
+    classes.
+    Called on every epoch.
+    :param model: pytorch model to be trained
+    :type model: torch.nn.Module
+    :param loader: DataLoader configured for the epoch.
+    :type loader: :class:`torch.utils.data.DataLoader`
+    :param optimizer: Optimizer object used to train the model.
+    :type optimizer: :class:`torch.optim.Optimizer`
+    :param criterion: loss function to use
+    :type criterion: function
+    :param active_classes: list of int specifying the "active" classes
+    :type active_classes: list or None
+    :param device:
+    :type device: :class:`torch.device`
+    :param post_batch_callback: function(model) to call after every batch
+    :type post_batch_callback: function
+    """
+    model.train()
+    for data, target in tqdm(loader, desc="Train", leave=False):
+        data, target = data.to(device), target.to(device)
+        if active_classes is not None:
+            target = target - active_classes[0]
```

Comment:
There is a similar transformation applied as a target_transform in `subdataset.get_datasets`. What is the difference, why is in that case applied directly to the dataset, while here is applied in the training function after the target is sampled from the dataloader?

Reply:
You're right, I placed all of it in the subdataset class.